core/src/wallet/ethereum/EthereumLikeAccount.cpp

```diff
@@ -480,43 +481,42 @@ namespace ledger {
-                            payload->putString(api::Account::EV_SYNC_ERROR_CODE, api::to_string(result.getFailure().getErrorCode()));
-                            payload->putInt(api::Account::EV_SYNC_ERROR_CODE_INT, (int32_t)result.getFailure().getErrorCode());
-                            payload->putString(api::Account::EV_SYNC_ERROR_MESSAGE, result.getFailure().getMessage());
-                    }
-                    eventPublisher->postSticky(std::make_shared<Event>(code, payload), 0);
-                    std::lock_guard<std::mutex> lock(self->_synchronizationLock);
-                    self->_currentSyncEventBus = nullptr;
+            std::lock_guard<std::mutex> lock(_synchronizationLock);
+            if (_currentSyncEventBus)
+                    return _currentSyncEventBus;
+            auto eventPublisher = std::make_shared<EventPublisher>(getContext());
+
+            _currentSyncEventBus = eventPublisher->getEventBus();
+            auto future = _synchronizer->synchronize(std::static_pointer_cast<EthereumLikeAccount>(shared_from_this()))->getFuture();
+            auto self = std::static_pointer_cast<EthereumLikeAccount>(shared_from_this());
+
+            //Update current block height (needed to compute trust level)
+            _explorer->getCurrentBlock().onComplete(getContext(), [self] (const TryPtr<EthereumLikeBlockchainExplorer::Block>& block) mutable {
+                if (block.isSuccess()) {
+                    self->_currentBlockHeight = block.getValue()->height;
+                }
+            });
 
-                });
-                return eventPublisher->getEventBus();
+            auto startTime = DateUtils::now();
+            eventPublisher->postSticky(std::make_shared<Event>(api::EventCode::SYNCHRONIZATION_STARTED, api::DynamicObject::newInstance()), 0);
+            future.onComplete(getContext(), [eventPublisher, self, startTime] (const Try<Unit>& result) {
+                api::EventCode code;
+                auto payload = std::make_shared<DynamicObject>();
+                auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(DateUtils::now() - startTime).count();
+                payload->putLong(api::Account::EV_SYNC_DURATION_MS, duration);
+                if (result.isSuccess()) {
+                    code = api::EventCode::SYNCHRONIZATION_SUCCEED;
+                } else {
+                    code = api::EventCode::SYNCHRONIZATION_FAILED;
+                    payload->putString(api::Account::EV_SYNC_ERROR_CODE, api::to_string(result.getFailure().getErrorCode()));
+                    payload->putInt(api::Account::EV_SYNC_ERROR_CODE_INT, (int32_t)result.getFailure().getErrorCode());
+                    payload->putString(api::Account::EV_SYNC_ERROR_MESSAGE, result.getFailure().getMessage());
+                }
+                eventPublisher->postSticky(std::make_shared<Event>(code, payload), 0);
+                std::lock_guard<std::mutex> lock(self->_synchronizationLock);
```

Comment:
As I said in the PR about TTL, there is probably a dead lock here. Please use a recursive mutex at least or find a solution to avoid the clashing

Reply:
So first, the only thing I changed about that code is the execution context in which we complete future, then I reformatted the code.
And by the way, I don't think there will be an issue since we call `onComplete` on a different thread.

Reply:
@alekece so `synchronize` will be called in the "global" execution context of the host, then when the future is completed, the rest of the code will be called on a thread of the execution context of the account.